File: legacy/python/src/features/groupadmin.py
"""Group admin feature: kick/ban/mute, anti-link, welcome message, bad words."""

import re

from aiogram import Bot, F, Router
from aiogram.filters import Command, ChatMemberUpdatedFilter, IS_MEMBER, IS_NOT_MEMBER
from aiogram.types import ChatMemberUpdated, ChatPermissions, Message

from .. import db

LINK_RE = re.compile(r"(https?://|t\.me/|www\.)", re.IGNORECASE)
BAD_WORDS = ["کص", "کیر", "کون", "جنده", "فحش"]


def build_router(tenant_id: int) -> Router:
    router = Router(name=f"ga-{tenant_id}")

    async def owner(message: Message) -> bool:
        t = await db.get_tenant(tenant_id)
        return bool(t and t["owner_id"] == message.from_user.id)

    async def is_admin(bot: Bot, chat_id: int, user_id: int) -> bool:
        try:
            member = await bot.get_chat_member(chat_id, user_id)
            return member.status in ("administrator", "creator")
        except Exception:
            return False

    async def target_user(message: Message) -> int | None:
        if message.reply_to_message:
            return message.reply_to_message.from_user.id if message.reply_to_message.from_user else None
        parts = message.text.split()
        if len(parts) >= 2 and parts[1].lstrip("@").isdigit():
            return int(parts[1].lstrip("@"))
        return None

    # ---- moderation commands (owner only) ----

    @router.message(Command("kick"))
    async def cmd_kick(message: Message):
        if not await owner(message):
            return
        uid = await target_user(message)
        if not uid:
            return await message.answer("روی پیام کسی ریپلای کن یا آیدی بده. مثال: /kick 12345")
        try:
            await message.chat.ban_chat_member(uid)
            await message.chat.unban_chat_member(uid)
            await message.answer("👢 کیک شد.")
        except Exception as e:
            await message.answer(f"خطا: {e}")

    @router.message(Command("ban"))
    async def cmd_ban(message: Message):
        if not await owner(message):
            return
        uid = await target_user(message)
        if not uid:
            return await message.answer("روی پیام کسی ریپلای کن یا آیدی بده. مثال: /ban 12345")
        try:
            await message.chat.ban_chat_member(uid)
            await message.answer("⛔️ بن شد.")
        except Exception as e:
            await message.answer(f"خطا: {e}")

    @router.message(Command("unban"))
    async def cmd_unban(message: Message):
        if not await owner(message):
            return
        uid = await target_user(message)
        if not uid:
            return await message.answer("مثال: /unban 12345")
        try:
            await message.chat.unban_chat_member(uid)
            await message.answer("✅ آنبن شد.")
        except Exception as e:
            await message.answer(f"خطا: {e}")

    @router.message(Command("mute"))
    async def cmd_mute(message: Message):
        if not await owner(message):
            return
        uid = await target_user(message)
        if not uid:
            return await message.answer("روی پیام کسی ریپلای کن یا آیدی بده. مثال: /mute 12345")
        try:
            await message.chat.restrict_chat_member(uid, ChatPermissions(can_send_messages=False))
            await message.answer("🔇 میوت شد.")
        except Exception as e:
            await message.answer(f"خطا: {e}")

    @router.message(Command("unmute"))
    async def cmd_unmute(message: Message):
        if not await owner(message):
            return
        uid = await target_user(message)
        if not uid:
            return await message.answer("مثال: /unmute 12345")
        try:
            await message.chat.restrict_chat_member(
                uid,
                ChatPermissions(
                    can_send_messages=True,
                    can_send_media_messages=True,
                    can_send_other_messages=True,
                    can_add_web_page_previews=True,
                ),
            )
            await message.answer("✅ آنمیوت شد.")
        except Exception as e:
            await message.answer(f"خطا: {e}")

    @router.message(Command("antilink"))
    async def cmd_antilink(message: Message):
        if not await owner(message):
            return
        cfg = await db.get_config(tenant_id)
        if not cfg["features"].get("antispam") and not cfg["features"].get("groupadmin"):
            return await message.answer("ربات تو این قابلیت رو نداره.")
        arg = message.text.replace("/antilink", "", 1).strip().lower()
        if arg not in ("on", "off"):
            return await message.answer("مثال: /antilink on یا /antilink off")
        cfg["antilink"] = arg == "on"
        await db.save_config(tenant_id, cfg)
        await message.answer(f"ضدلینک: {'✅ فعال' if cfg['antilink'] else '❌ غیرفعال'}")

    @router.message(Command("setwelcome"))
    async def cmd_setwelcome(message: Message):
        if not await owner(message):
            return
        cfg = await db.get_config(tenant_id)
        if not cfg["features"].get("welcome") and not cfg["features"].get("groupadmin"):
            return await message.answer("ربات تو این قابلیت رو نداره.")
        text = message.text.replace("/setwelcome", "", 1).strip()
        if not text:
            return await message.answer("مثال: /setwelcome خوش اومدی {name} عزیز! 🎉")
        cfg["welcome_msg"] = text
        await db.save_config(tenant_id, cfg)
        await message.answer("✅ پیام خوش‌آمد ذخیره شد. از {name} برای اسم کاربر استفاده کن.")

    @router.message(Command("welcome"))
    async def cmd_welcome_preview(message: Message):
        if not await owner(message):
            return
        cfg = await db.get_config(tenant_id)
        await message.answer(
            (cfg.get("welcome_msg") or "خوش اومدی {name}!").replace("{name}", message.from_user.first_name or "")
        )

    # ---- automatic group protection ----

    @router.message(F.chat.type.in_({"group", "supergroup"}), F.text)
    async def group_policer(message: Message):
        # skip owner and admins
        if await owner(message) or await is_admin(message.bot, message.chat.id, message.from_user.id):
            return
        cfg = await db.get_config(tenant_id)
        can_act = cfg["features"].get("antispam") or cfg["features"].get("groupadmin")

        # anti-link
        if can_act and cfg.get("antilink") and LINK_RE.search(message.text or ""):
            try:
                await message.delete()
                await message.answer(f"⚠️ {message.from_user.first_name}، لینک در این گروه ممنوعه!", disable_notification=True)
            except Exception:
                pass
            return

        # bad words
        if can_act and cfg.get("antispam"):
            for w in BAD_WORDS:
                if w in (message.text or "").lower():
                    try:
                        await message.delete()
                    except Exception:
                        pass
                    return

    @router.chat_member(ChatMemberUpdatedFilter(member_status_changed=(IS_NOT_MEMBER, IS_MEMBER)))
    async def on_new_member(event: ChatMemberUpdated):
        if event.new_chat_member.status != "member":
            return
        cfg = await db.get_config(tenant_id)
        if not (cfg.get("welcome_msg") or "").strip():
            return
        if not (cfg["features"].get("welcome") or cfg["features"].get("groupadmin")):
            return
        name = event.new_chat_member.user.first_name or "دوست عزیز"
        try:
            await event.answer(cfg["welcome_msg"].replace("{name}", name))
        except Exception:
            pass

    return router

File: src/features/groupadmin.ts
/** Feature: group admin — kick/ban/mute, antilink, bad-words filter, welcome message.
 *  Covers the `groupadmin`, `antispam` and `welcome` templates. */

import type { Bot, Context } from "grammy";

import { hasLink } from "../templates";
import type { TenantCtx } from "../panel";

function arg(ctx: Context): string {
  return typeof ctx.match === "string" ? ctx.match.trim() : "";
}

async function isAdmin(bot: Bot, chatId: number, userId: number): Promise<boolean> {
  try {
    const m = await bot.api.getChatMember(chatId, userId);
    return ["administrator", "creator"].includes(m.status);
  } catch {
    return false;
  }
}

export function registerGroupAdmin(bot: Bot, t: TenantCtx): void {
  const hasGroupAdmin = t.feature("groupadmin");
  const hasAntispam = t.feature("antispam");
  const hasWelcome = t.feature("welcome");
  if (!hasGroupAdmin && !hasAntispam && !hasWelcome) return;

  if (!t.config.badWords) t.config.badWords = [];

  // ---------- moderation (groupadmin template) ----------
  if (hasGroupAdmin) {
    bot.command("kick", async (ctx) => {
      if (!ctx.from || !ctx.message) return;
      if (!(await isAdmin(bot, ctx.chat.id, ctx.from.id))) return;
      const target = ctx.message.reply_to_message?.from?.id;
      if (!target) return ctx.reply("روی پیام کسی ریپلای کن.");
      try {
        await bot.api.banChatMember(ctx.chat.id, target);
        await bot.api.unbanChatMember(ctx.chat.id, target);
        await ctx.reply("👢 کیک شد.");
      } catch {
        await ctx.reply("ربات ادمین نیست یا خطایی رخ داد.");
      }
    });

    bot.command("ban", async (ctx) => {
      if (!ctx.from || !ctx.message) return;
      if (!(await isAdmin(bot, ctx.chat.id, ctx.from.id))) return;
      const target = ctx.message.reply_to_message?.from?.id;
      if (!target) return ctx.reply("روی پیام کسی ریپلای کن.");
      try {
        await bot.api.banChatMember(ctx.chat.id, target);
        await ctx.reply("🚫 بن شد.");
      } catch {
        await ctx.reply("ربات ادمین نیست یا خطایی رخ داد.");
      }
    });

    bot.command("unban", async (ctx) => {
      if (!ctx.from) return;
      if (!(await isAdmin(bot, ctx.chat.id, ctx.from.id))) return;
      const parts = arg(ctx).split(/\s+/);
      const id = parts[0] && /^\d+$/.test(parts[0]) ? Number(parts[0]) : null;
      if (!id) return ctx.reply("مثال: /unban 123456789");
      try {
        await bot.api.unbanChatMember(ctx.chat.id, id);
        await ctx.reply("✅ آنبن شد.");
      } catch {
        await ctx.reply("خطایی رخ داد.");
      }
    });

    bot.command("mute", async (ctx) => {
      if (!ctx.from || !ctx.message) return;
      if (!(await isAdmin(bot, ctx.chat.id, ctx.from.id))) return;
      const target = ctx.message.reply_to_message?.from?.id;
      if (!target) return ctx.reply("روی پیام کسی ریپلای کن.");
      try {
        await bot.api.restrictChatMember(ctx.chat.id, target, { can_send_messages: false });
        await ctx.reply("🔇 میوت شد.");
      } catch {
        await ctx.reply("ربات ادمین نیست یا خطایی رخ داد.");
      }
    });

    bot.command("unmute", async (ctx) => {
      if (!ctx.from || !ctx.message) return;
      if (!(await isAdmin(bot, ctx.chat.id, ctx.from.id))) return;
      const target = ctx.message.reply_to_message?.from?.id;
      if (!target) return ctx.reply("روی پیام کسی ریپلای کن.");
      try {
        await bot.api.restrictChatMember(ctx.chat.id, target, { can_send_messages: true });
        await ctx.reply("🔊 آنمیوت شد.");
      } catch {
        await ctx.reply("ربات ادمین نیست یا خطایی رخ داد.");
      }
    });
  }

  // ---------- channel posting (groupadmin template) ----------
  if (hasGroupAdmin) {
    bot.command("setchannel", async (ctx) => {
      if (!ctx.from || !t.isOwner(ctx.from.id)) return;
      const raw = arg(ctx);
      if (!raw) return ctx.reply("مثال: /setchannel @mychannel");
      const target = /^\d+$/.test(raw) ? Number(raw) : raw;
      try {
        const chat = await bot.api.getChat(target);
        t.config.channel = chat.id;
        await t.save(bot);
        await ctx.reply(
          `✅ کانال «${chat.title ?? chat.username ?? chat.id}» تنظیم شد.\n` +
            `حالا /post <متن> بفرست تا توش پست بذاری.`
        );
      } catch {
        await ctx.reply("کانال پیدا نشد — مطمئن شو ربات ادمینشه و @ یا آیدی عددی بفرست.");
      }
    });

    bot.command("post", async (ctx) => {
      if (!ctx.from || !t.isOwner(ctx.from.id)) return;
      const text = arg(ctx);
      if (!text) return ctx.reply("مثال: /post سلام دنیا 🚀");
      const ch = t.config.channel;
      if (!ch) return ctx.reply("اول /setchannel @channel رو بزن.");
      try {
        await bot.api.sendMessage(ch, text);
        await ctx.reply("✅ پست توی کانال ارسال شد.");
      } catch {
        await ctx.reply("ارسال نشد — ربات ادمین کاناله؟");
      }
    });
  }

  // ---------- antilink + bad words (groupadmin / antispam templates) ----------
  if (hasGroupAdmin || hasAntispam) {
    bot.command("antilink", async (ctx) => {
      if (!ctx.from || !t.isOwner(ctx.from.id)) return;
      const v = arg(ctx).toLowerCase();
      if (v !== "on" && v !== "off") return ctx.reply("مثال: /antilink on  یا  /antilink off");
      t.config.antilink = v === "on";
      await t.save(bot);
      await ctx.reply(`✅ آنتی‌لینک: ${v === "on" ? "روشن" : "خاموش"}`);
    });

    bot.command("addword", async (ctx) => {
      if (!ctx.from || !t.isOwner(ctx.from.id)) return;
      const w = arg(ctx);
      if (!w) return ctx.reply("مثال: /addword فحش");
      if (!t.config.badWords!.includes(w)) t.config.badWords!.push(w);
      await t.save(bot);
      await ctx.reply(`✅ کلمه «${w}» به لیست اضافه شد.`);
    });

    bot.on("message", async (ctx, next) => {
      const chat = ctx.chat;
      if (chat.type !== "group" && chat.type !== "supergroup") return next();
      if (!ctx.message?.text) return next();
      const sender = ctx.message.from;
      if (!sender || (await isAdmin(bot, chat.id, sender.id))) return next();

      const text = ctx.message.text;
      let bad = false;
      if (t.config.antilink && hasLink(text)) bad = true;
      if (!bad && t.config.badWords!.some((w) => text.includes(w))) bad = true;

      if (bad) {
        try {
          await bot.api.deleteMessage(chat.id, ctx.message.message_id);
          const warn = await bot.api.sendMessage(chat.id, `🚫 پیام ${sender.first_name} حذف شد (ضداسپم).`);
          setTimeout(() => bot.api.deleteMessage(chat.id, warn.message_id).catch(() => {}), 5000);
        } catch {
          // bot not admin
        }
      }
      return next();
    });
  }

  // ---------- welcome (groupadmin / welcome templates) ----------
  if (hasGroupAdmin || hasWelcome) {
    bot.command("setwelcome", async (ctx) => {
      if (!ctx.from || !t.isOwner(ctx.from.id)) return;
      const text = arg(ctx);
      if (!text) return ctx.reply("مثال: /setwelcome خوش اومدی {name} 👋");
      t.config.welcomeText = text;
      await t.save(bot);
      await ctx.reply("✅ متن خوش‌آمد تنظیم شد.");
    });

    bot.on("chat_member", async (ctx) => {
      const u = ctx.update.chat_member;
      const becameMember =
        u.new_chat_member.status === "member" &&
        u.old_chat_member.status !== "member" &&
        u.old_chat_member.status !== "administrator";
      if (!becameMember) return;
      if (!t.config.welcomeText) return;
      const name = u.new_chat_member.user.first_name;
      try {
        await ctx.api.sendMessage(u.chat.id, t.config.welcomeText.replace(/\{name\}/g, name));
      } catch {
        // bot not admin / chat gone
      }
    });
  }
}
